### ISSUE TITLE
Node de-duplication for network layer?

### ISSUE THREAD

**Issue by tuukka** [facebook/relay#806]
Would it make sense to implement node de-duplication via a new serialisation format between GraphQL servers and Relay? There's an open issue on GraphQL's side, but perhaps the spec is fine as it already allows results other than plain JSON: https://github.com/facebook/graphql/issues/130 vs http://facebook.github.io/graphql/#sec-Serialization-Format

Any musings on the design space?

1. HTTP requests from Relay announce support via an Accept header / a query parameter?
2. GraphQL server renders the de-duped response as JSON Graph / JSON-LD / JSON (Hyper-)Schema / Protobuf / ... ?
3. In Relay, the network layer stores the new nodes in cache and passes on the spine of the response? How to save most time in JSON manipulation and get directly to React props?

**Comment by josephsavona** [comment]
This is something we've explored and may implement going forward. The performance benefits - especially in terms of network bytes - are not necessarily what you'd expect, however. @wincent did some analysis here and can probably share more.

**Comment by wincent** [comment]
I did a some experimentation on this for a hackathon and wrote up the results in a fair amount of detail [here](https://github.com/wincent/relay/tree/hack/contrib/lightweight-results).

The (possibly counterintuitive) result is that no combination or normalization, binary encoding, and compression beats naive, fully-expanded JSON compressed with GZIP. GZIP is _really_ good at eliminating redundancy, and as I say in my write-up:

> In short, if you try to eliminate redundancy by applying some kind of clever transformation or restructuring your data object before piping it to gzip, you will probably just do a crappy job of reimplementing gzip at the wrong layer of abstraction, and get worse results overall.

Having said that, all hope is not lost. I listed some ideas under [the "Future Work" section](https://github.com/wincent/relay/tree/hack/contrib/lightweight-results#conclusions-and-future-work) of the write-up:

- Better align network format with what operations the Relay store uses during writes (won't be smaller over the network, but the writes could be made cheaper/faster).
- Relatedly, use a stream-ier network transport to enable beginning writing/processing of the data as soon as the first bytes get flushed and arrive back on the client.
- Keep a synchronized proxy of the client cache on the server side so that it can compute delta updates instead of full payloads (ie. avoid sending objects/fields which it knows to already exist on the client).

Unfortunately, all of these ideas require significant changes that span the client and server divide.

Thanks for bringing this up, @tuukka. I am going to close this for now as there's nothing immediately actionable in here for now. (It will probably be a long time before we get to some of the more ambitious/experimental ideas, but when we do we should create separate issues for them.) In the meantime if you wanted to share a hack or a proof-of-concept for any of these, we'd love to see it in the form of a PR (even if only a "throwaway" one).
